contracts/utils/MultiRateLimited.sol

```diff
@@ -0,0 +1,273 @@
+        require(newBufferCap <= bufferCap, "MultiRateLimited: new buffer cap is over global max");
+
+        _setIndividualBufferCap(rateLimitedAddress, newBufferCap);
+    }
+
+    // ----------- Getters -----------
+
+    /// @notice the amount of action used before hitting limit
+    /// @dev replenishes at rateLimitPerSecond per second up to bufferCap
+    /// @param rateLimitedAddress the address whose buffer will be returned
+    function individualBuffer(address rateLimitedAddress) public view override returns(uint144) {
+        RateLimitData memory rateLimitData = rateLimitPerAddress[rateLimitedAddress];
+
+        uint256 elapsed = block.timestamp - rateLimitData.lastBufferUsedTime;
+        return uint144(Math.min(rateLimitData.bufferStored + (rateLimitData.rateLimitPerSecond * elapsed), rateLimitData.bufferCap));
+    }
+
+    /// @notice the rate per second for each address
+    function getRateLimitPerSecond(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].rateLimitPerSecond;
+    }
+
+    /// @notice the last time the buffer was used by each address
+    function getLastBufferUsedTime(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].lastBufferUsedTime;
+    }
+
+    /// @notice the cap of the buffer that can be used at once
+    function getBufferCap(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].bufferCap;
+    }
+
+    // ----------- Helper Methods -----------
+
+    function _updateAddress(address rateLimitedAddress, uint112 _rateLimitPerSecond, uint144 _bufferCap) internal {
+        uint112 oldRateLimitPerSecond = _rateLimitPerSecond;
+
+        RateLimitData storage rateLimitData = rateLimitPerAddress[rateLimitedAddress];
+
+        require(rateLimitData.lastBufferUsedTime != 0, "MultiRateLimited: rate limit address does not exist");
+        require(_rateLimitPerSecond <= MAX_RATE_LIMIT_PER_SECOND, "MultiRateLimited: rateLimitPerSecond too high");
+
+        rateLimitData.lastBufferUsedTime = block.timestamp.toUint32();
+        rateLimitData.bufferCap = _bufferCap;
+        rateLimitData.rateLimitPerSecond = _rateLimitPerSecond;
+        rateLimitData.bufferStored = _bufferCap;
+
+        emit IndividualRateLimitPerSecondUpdate(rateLimitedAddress, oldRateLimitPerSecond, _rateLimitPerSecond);
+    }
+
+    /// @param rateLimitedAddress the new address to add as a rateLimitedAddress
+    /// @param _rateLimitPerSecond the rate limit per second for this rateLimitedAddress
+    /// @param _bufferCap  the buffer cap for this rateLimitedAddress
+    function _addAddress(address rateLimitedAddress, uint112 _rateLimitPerSecond, uint144 _bufferCap) internal {
+        require(_bufferCap <= bufferCap, "MultiRateLimited: new buffercap too high");
```

Comment:
Took me awhile to realize that `bufferCap` is coming from the base RateLimited contract. I wish there were a way to alias this to `globalBufferCap` or something. Can you add a comment explaining the origin of this variable, and that it's the global one?

Reply:
(closing, will fix in my nitpicks pr)